### DISCUSSION TITLE
routes appear under tag of Apirouter and parent Apirouter

### DISCUSSION THREAD

**Discussion by transfluxus** [fastapi/fastapi#7097] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [ ] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [ ] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [ ] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [ ] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [ ] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
import uvicorn
from fastapi import FastAPI, APIRouter

items = {}
app = FastAPI()

router = APIRouter(prefix="/api")
items_router = APIRouter(prefix="/items", tags=["items"])

@items_router.post("/{id}")
def add_item(id: str):
	items[id] = {"attachments": []}
	return items[id]

@items_router.get("/{id}")
def get_item(id: str):
	if not (item := items.get(id)):
		return "error"
	return item

items_attachment_router = APIRouter(prefix="/{id}", tags=["attachment"])

@items_attachment_router.post("/add")
def add_attachment(id: str, attachment: str):
	if not (item := items.get(id)):
		return "error"
	item["attachments"].append(attachment)
	return item

items_attachment_router2 = APIRouter()

@items_attachment_router2.post("/{id}/add2")
def add_attachment2(id: str, attachment: str):
	if not (item := items.get(id)):
		return "error"
	item["attachments"].append(attachment)
	return item

items_router.include_router(items_attachment_router)
items_router.include_router(items_attachment_router2, tags=["attachment2"])
router.include_router(items_router)
app.include_router(router)

if __name__ == "__main__":
	uvicorn.run(app)

```

### Description

the routes of an APIRouter appear under the tags given by the router and all parent routers.
I implemented it the old and the new way (adding tags to the router directly & adding them as a param in include_router). Both methods put the routes add_attachment and add_attachment2 under their own tag and them items tag
* Python version: 3.8

### Additional context
FastAPI version 0.62.0

**Comment by Mause** [comment]
Sorry, can you explain what the issue is? It isn't clear from the information you have provided

**Comment by ajju10** [comment]
Please describe the issue clearly or better to share a screenshot of the page you are getting issue with along with the clean description of your issue.

**Comment by transfluxus** [comment]
Hi,

sorry, the issue is that the routes appear twice (in swagger)

https://imagebin.ca/v/5kgpOV3O3InG

**Comment by Mause** [comment]
This is not a bug in fastapi, please see https://github.com/swagger-api/swagger-ui/issues/3711

**Comment by transfluxus** [comment]
Hi @Mause ,
thanks for your answer. In the thread you send multiple tags are added to the endpoints. In my case the motivation is to have a large API better structured.
Different modules, different ApiRouter objects, different tags. Still using hierarchies of routers for dependency purposes.
So routes in multiple locations doesnt help a lot to have the swagger page better sorted :)

So I see, after calling  

I dont see exactly where fastapi puts these tags together.
in routing `include_router` there is
```
                current_tags = []
                if tags:
                    current_tags.extend(tags)
                if route.tags:
                    current_tags.extend(route.tags)
```
but going through it with the debugger, there is no moment when I see the tags: _items_ and _attachment_ being put together.
Only at the point I call  `router.include_router(items_router)`  `route.tags` has both of them

So in order to fix this (for my app), any better way how I can change the tags of the routes than this:
```

...
items_router.include_router(items_attachment_router)

for route in items_router.routes:
	if route.name in [r.name for r in items_attachment_router.routes]:
		for tag in items_router.tags:
			route.tags.remove(tag)
```

Maybe a parameter to the `include_router` could avoid this hack

**Comment by Mause** [comment]
I'm afraid I don't understand what you are trying to achieve - if you don't want a tag to be added, don't add it in the first place?

**Comment by transfluxus** [comment]
see what the hack made:

https://imagebin.ca/v/5lIfxYIs3VRX compared to the 1. one.

**Comment by transfluxus** [comment]
while still having the hierarchy of routers, which makes a bunch of things convinient

**Comment by ycd** [comment]
@transfluxus isn't it the expected behavior? 

In the end, all of the routers extend your actual application and its OpenAPI spec.

**Comment by transfluxus** [comment]
... sure but this is about tags, and tags being passed down from one router to another...

**Comment by Mause** [comment]
Why does this being about tags change anything? Everything you have described so far is expected behaviour

**Comment by transfluxus** [comment]
k, what you would expect...

Its aruably reasonable to not have all routes in the api doc multiple times and only found a hack to to solve it.

So this is a feature request

**Comment by Mause** [comment]
If you do not wish for routes to appear multiple times in the Swagger UI, you must only have one tag per route. We do not control this behaviour, this is functionality built into Swagger UI, as I stated above

**Comment by transfluxus** [comment]
I have only one tag per router as stated and shown above. I found a way to change the include_router and add_route method to change the tags, which could be a small pull request but I'll leave it. This behaviour comes basically from the fact that add_router basically just throws the routes for the passed router into its existing ones, adding the tags of **both routers**. Considering that, I will just not chain my routers but just include them separately to the main router.

**Comment by tiangolo** [comment, marked as answer]
Thanks for the help here everyone! :clap: :bow: 

Thanks for reporting back and closing the issue :+1:

More info just for completeness, in terms of the OpenAPI spec, the tags are just a list of tags for related path operations. The fact that Swagger UI shows them grouped by tag is a specific behavior and/or feature of Swagger UI. And depending on the use case, having the same path operation showing under different groups (having multiple tags) could be seen as a "feature". The OpenAPI spec doesn't enforce how tags should be used. So, it would make sense to merge tags if they are set in multiple places.

In your case, you found the solution :rocket: To separate routes that should not be combined under the same tags.

If for some reason you still wanted to include one sub-router in another (say, to share common dependencies) as in:

```Python
items_router.include_router(items_attachment_router)
```

...what you could do is have a parent router with all the shared parts (e.g. dependencies), e.g. `items_router`. And instead of adding path operations with that parent router, you could have another sub-router with just the tag e.g. `items_router_paths`, and add the "top-level" path operations to that one. Something like:

```Python
items_router.include_router(items_router_path)
items_router.include_router(items_attachment_router)
```

**Comment by joaovbsevero** [comment]
Just like to add in here that the "solution" defeats the purpose of having hierarchical routers: 

FastAPI is awesome 💯  by allowing us to compose applications in a simple way by splitting it into sub-modules and constructing a hierarchy of routers 

Although, once we take advantage of that feature we polute our documentation and the solution for that is to remove route composition 🫠   

Since the router composition cannot be used, I would argue to remove the possibility to have router composition in the first place since FastAPI will not be able to provide a clean interface for a properly documented hierarchy. 

And, as Mause insisted 🙄 , since it is an issue with Swagger, why not get rid of router composition to respect Swagger? Or maybe there is a better solution.. 😄
